website/profile/views.py

```diff
@@ -85,20 +87,89 @@ def update_user(auth):
     data = request.get_json()
 
     # TODO: Expand this to support other user attributes
-    if 'timezone' in data:
-        if data['timezone']:
-            user.timezone = data['timezone']
+
+    ##########
+    # Emails #
+    ##########
+
+    if 'emails' in data:
+        # removals
+        removed_emails = [
+            each
+            for each in user.emails + user.unconfirmed_emails
+            if each not in [x['address'].strip().lower()
+                            for x in data['emails']]
+        ]
+
+        for address in removed_emails:
+            if address in user.emails:
+                user.remove_email(address)
+            user.remove_unconfirmed_email(address)
+
+        # additions
+        added_emails = [
+            each['address'].strip().lower()
+            for each in data['emails']
+            if each['address'].strip().lower() not in user.emails
+            and each['address'].strip().lower() not in user.unconfirmed_emails
+        ]
+
+        for address in added_emails:
+            try:
+                user.add_unconfirmed_email(address)
+            except ValidationError:
+                continue
+
+            # TODO: This setting is now named incorrectly.
+            if settings.CONFIRM_REGISTRATIONS_BY_EMAIL:
+                send_confirm_email(user, email=address)
+
+        ############
+        # Username #
+        ############
+
+        # get the first email that is set to primary and has an address
+        primary_email = next(
+            (
+                each for each in data['emails']
+                # email is primary
+                if each.get('primary')
+                # an address is specified (can't trust those sneaky users!)
+                and each.get('address')
+            )
+        )
+
+        if primary_email:
+            username = primary_email['address'].strip().lower()
+
+        # make sure the new username has already been confirmed
+        if username and username in user.emails and username != user.username:
+            mails.send_mail(user.username,
+                            mails.PRIMARY_EMAIL_CHANGED,
+                            user=user,
+                            new_address=username)
+            user.username = username
```

Comment:
I wonder if it would make sense to have a `set_primary_email` method on User which

- Preprocesses email (lowercas and strip)
- Sets username field
- Emits a signal

Another method could subscribe to the signal and send the email.